File: app/src/main/java/net/squanchy/home/deeplink/HomeActivityIntentParser.java
package net.squanchy.home.deeplink;

import android.content.Intent;
import android.os.Bundle;

import net.squanchy.home.BottomNavigationSection;
import net.squanchy.home.HomeStatePersister;
import net.squanchy.support.lang.Optional;

import static net.squanchy.home.deeplink.HomeActivityDeepLinkCreator.KEY_INITIAL_PAGE_INDEX;

public class HomeActivityIntentParser {

    private final HomeStatePersister statePersister;
    private final Optional<Bundle> savedState;
    private final Intent intent;

    public HomeActivityIntentParser(Optional<Bundle> savedState, Intent intent) {
        this.statePersister = new HomeStatePersister();
        this.savedState = savedState;
        this.intent = intent;
    }

    public BottomNavigationSection getInitialSelectedPage() {
        return getSectionFrom(statePersister, savedState)
                .or(getSectionFrom(Optional.fromNullable(intent.getExtras())))
                .or(BottomNavigationSection.SCHEDULE);
    }

    private Optional<BottomNavigationSection> getSectionFrom(HomeStatePersister statePersister, Optional<Bundle> savedState) {
        return statePersister.readSectionFrom(savedState);
    }

    private Optional<BottomNavigationSection> getSectionFrom(Optional<Bundle> intentExtras) {
        Bundle bundle = intentExtras.or(Bundle.EMPTY);

        if (bundle.containsKey(KEY_INITIAL_PAGE_INDEX)) {
            int sectionIndex = bundle.getInt(KEY_INITIAL_PAGE_INDEX);
            return Optional.of(BottomNavigationSection.values()[sectionIndex]);
        } else {
            return Optional.absent();
        }
    }
}

File: app/src/main/java/net/squanchy/home/deeplink/HomeActivityIntentParser.kt
package net.squanchy.home.deeplink

import android.content.Intent
import android.os.Bundle

import net.squanchy.home.BottomNavigationSection
import net.squanchy.home.HomeStatePersister

import net.squanchy.home.deeplink.HomeActivityDeepLinkCreator.KEY_INITIAL_PAGE_INDEX

class HomeActivityIntentParser(private val savedState: Bundle?, private val intent: Intent) {

    private val statePersister: HomeStatePersister = HomeStatePersister()

    val initialSelectedPage: BottomNavigationSection
        get() = getSectionFrom(statePersister, savedState)
                ?: getSectionFrom(intent.extras)
                ?: BottomNavigationSection.SCHEDULE

    private fun getSectionFrom(statePersister: HomeStatePersister, savedState: Bundle?): BottomNavigationSection? {
        return statePersister.readSectionFrom(savedState)
    }

    private fun getSectionFrom(intentExtras: Bundle?): BottomNavigationSection? {
        val bundle = intentExtras ?: Bundle.EMPTY

        if (bundle.containsKey(KEY_INITIAL_PAGE_INDEX)) {
            val sectionIndex = bundle.getInt(KEY_INITIAL_PAGE_INDEX)
            return BottomNavigationSection.values()[sectionIndex]
        } else {
            return null
        }
    }
}
